File: sdks/python/plexspaces/ws_thin_client.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (C) 2025 PlexSpaces Contributors
#
# WsThinClient — Python thin-node WebSocket client skeleton.
#
# TODO: Implement WsFrame protobuf encoding over the `websockets` library.
# Wire protocol: proto/plexspaces/v1/transport/websocket.proto
# Reference implementation: sdks/typescript/src/ws_thin_client.ts

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, Optional


@dataclass
class WsThinClientOptions:
    """Options for connecting a thin node via WebSocket."""
    ws_url: str
    """WebSocket URL, e.g. "ws://localhost:8091/ws"."""
    jwt_token: Optional[str] = None
    """JWT Bearer token. Appended as ?token=<jwt> if present."""
    node_id: Optional[str] = None
    """ULID preferred. Server assigns one if omitted or collision detected."""
    tenant: str = "default"
    namespace: str = "default"


@dataclass
class ThinNodePingResult:
    """Resource hints from a PingResponse (proto fields 9–11 of PingResponse)."""
    node_id: str
    cpu_percent: float
    memory_available_mb: int
    available_cores: int


class WsThinClient:
    """Python thin-node WebSocket client skeleton.

    All methods raise ``NotImplementedError``.

    TODO: implement WsFrame protobuf encoding over the ``websockets`` library.
    Reference: sdks/typescript/src/ws_thin_client.ts
    Proto: proto/plexspaces/v1/transport/websocket.proto
    """

    def __init__(self, opts: WsThinClientOptions) -> None:
        raise NotImplementedError(
            "WsThinClient: implement WsFrame protobuf encoding over websockets library"
        )

    async def connect(self) -> str:
        """Open the WebSocket and complete the NodeRegistration handshake.

        Returns the server-assigned node_id.
        """
        ...

    async def tell(self, actor_id: str, msg_type: str, payload: dict[str, Any]) -> None:
        """Fire-and-forget tell to a canonical actor ID."""
        ...

    async def ask(
        self,
        actor_id: str,
        msg_type: str,
        payload: dict[str, Any],
        timeout_ms: int = 5000,
    ) -> dict[str, Any]:
        """Request-reply ask. Returns the response payload."""
        ...

    def on_message(self, handler: Callable[[str, str, Any], None]) -> None:
        """Register a handler for incoming tell frames addressed to this thin node."""
        ...

    def node_id(self) -> str:
        """The server-assigned node_id (available after connect())."""
        ...

    def local_actor_id(self, name: str, actor_type: str, namespace: Optional[str] = None) -> str:
        """Build a canonical actor ID on this thin node.

        Format: {name}//{type}::{namespace}@{nodeId}
        """
        ns = namespace or "default"
        return f"{name}//{actor_type}::{ns}@{self.node_id()}"

    async def ping_node(self, target_node_id: str) -> ThinNodePingResult:
        """Send a SWIM-compatible ping and return resource hints from the response."""
        ...

    async def heartbeat(self) -> None:
        """Send a heartbeat frame to keep the WS session alive."""
        ...

    async def disconnect(self) -> None:
        """Close the WebSocket connection."""
        ...

File: sdks/go/plexspaces/ws_thin_client.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2025 PlexSpaces Contributors
//
// WsThinClient — Go thin-node WebSocket client skeleton.
//
// TODO: Implement WsFrame protobuf encoding over gorilla/websocket or
//       the standard library net/http WebSocket (Go 1.24+).
//
// Wire protocol: proto/plexspaces/v1/transport/websocket.proto
// Reference implementation: sdks/typescript/src/ws_thin_client.ts

package plexspaces

import (
	"context"
	"errors"
)

// WsThinClientOptions configures a thin-node WebSocket connection.
type WsThinClientOptions struct {
	// WebSocket URL, e.g. "ws://localhost:8091/ws"
	WsURL string
	// JWT Bearer token. Appended as ?token=<jwt> if non-empty.
	JwtToken string
	// ULID preferred. Server assigns one if empty or collision detected.
	NodeID    string
	Tenant    string
	Namespace string
}

// ThinNodePingResult carries resource hints from a PingResponse.
// Corresponds to proto fields 9–11 of PingResponse in node.proto.
type ThinNodePingResult struct {
	NodeID            string
	CPUPercent        float32
	MemoryAvailableMB uint64
	AvailableCores    uint32
}

// WsThinClient is a thin-node WebSocket client.
// All methods return errors until the TODO implementation is complete.
type WsThinClient struct{}

// NewWsThinClient returns a WsThinClient configured with opts.
func NewWsThinClient(_ WsThinClientOptions) *WsThinClient {
	return &WsThinClient{}
}

// Connect opens the WebSocket and completes the NodeRegistration handshake.
// Returns the server-assigned node_id on success.
func (c *WsThinClient) Connect(_ context.Context) (string, error) {
	return "", errors.New("TODO: implement WsFrame protobuf encoding over WebSocket")
}

// Tell sends a fire-and-forget message to a canonical actor ID.
func (c *WsThinClient) Tell(_, _ string, _ any) error {
	return errors.New("TODO")
}

// Ask sends a request-reply message and returns the response payload.
func (c *WsThinClient) Ask(_ context.Context, _, _ string, _ any) (any, error) {
	return nil, errors.New("TODO")
}

// OnMessage registers a handler for incoming tell frames addressed to this thin node.
func (c *WsThinClient) OnMessage(_ func(actorID, msgType string, payload any)) {}

// NodeID returns the server-assigned node_id (available after Connect).
func (c *WsThinClient) NodeID() string { return "" }

// LocalActorID builds a canonical actor ID on this thin node.
// Format: {name}//{type}::{namespace}@{nodeId}
func (c *WsThinClient) LocalActorID(name, actorType, namespace string) string {
	return name + "//" + actorType + "::" + namespace + "@" + c.NodeID()
}

// PingNode sends a SWIM-compatible ping and returns resource hints from the response.
func (c *WsThinClient) PingNode(_ context.Context, _ string) (ThinNodePingResult, error) {
	return ThinNodePingResult{}, errors.New("TODO")
}

// Heartbeat sends a heartbeat frame to keep the WS session alive.
func (c *WsThinClient) Heartbeat() error { return errors.New("TODO") }

// Disconnect closes the WebSocket connection.
func (c *WsThinClient) Disconnect() error { return errors.New("TODO") }
